R/carehomes_ifr_t.R

```diff
@@ -0,0 +1,364 @@
+                 length(step)))
+  }
+  if (nrow(I_weighted) != ncol(p$rel_susceptibility) * p$n_groups) {
+    stop(sprintf(
+      "Expected 'I_weighted' to have %d rows = %d groups x %d vacc classes",
+      p$n_groups * ncol(p$rel_susceptibility),
+      p$n_groups,
+      ncol(p$rel_susceptibility)))
+  }
+  if (ncol(I_weighted) != length(step)) {
+    stop(sprintf("Expected 'I_weighted' to have %d columns, following 'step'",
+                 length(step)))
+  }
+
+  IFR_by_group_and_vacc_class <-
+    carehomes_IFR_t_by_group_and_vacc_class(step, p)
+  beta <- sircovid_parameters_beta_expand(step, p$beta_step)
+  n_vacc_classes <- ncol(p$rel_susceptibility)
+
+
+  calculate_expected_infections <- function(t, no_vacc) {
+    m <- p$m
+    ages <- seq_len(p$n_age_groups)
+    ch <- seq(to = p$n_groups, length.out = 2)
+    m[ages, ] <- beta[t] * m[ages, ]
+    m[ch, ages] <- beta[t] * m[ch, ages]
+
+    m_extended <- matrix(t(matrix(m, p$n_groups, p$n_groups * n_vacc_classes)),
+                         p$n_groups * n_vacc_classes,
+                         p$n_groups * n_vacc_classes,
+                         byrow = TRUE)
+
+    ## probability of infection in next time step by group and vaccine class
+    if (no_vacc) {
+      ## exclude vaccine effects
+      prob_infection <- (1 - exp(-p$dt * (m_extended %*% I_weighted[, t])))
+    } else {
+      ## include vaccine effects
+      prob_infection <- (1 - exp(-p$dt * c(p$rel_susceptibility) *
+                     (m_extended %*% (I_weighted[, t] * c(p$rel_infectivity)))))
+    }
+
+    ## expected infections in next time step by group and vaccine class
+    expected_infections <-  S[, t] * prob_infection
+
+    expected_infections
+  }
+
+
+  calculate_weighted_ratio <- function(t, expected_infections,
+                                       drop_carehomes, no_vacc, type) {
+
+    ## Care home workers (CHW) and residents (CHR) in last two rows
+    ## and columns, remove for each vaccine class
+    if (drop_carehomes) {
+      i_CHR <- seq(p$n_groups, dim(S)[1], by = p$n_groups)
+      i_CHW <- i_CHR - 1
+      i_keep <- seq_len(dim(S)[1])[-c(i_CHW, i_CHR)]
+    } else {
+      i_keep <- seq_len(dim(S)[1])
+    }
+
+    if (no_vacc) {
+      ## same IFR by group across all vaccine classes
+      IFR_vec <- rep(c(IFR_by_group_and_vacc_class[[type]][, 1, t]),
+                     n_vacc_classes)
+    } else {
+      IFR_vec <- c(IFR_by_group_and_vacc_class[[type]][, , t])
+    }
+
+    weighted.mean(IFR_vec[i_keep], expected_infections[i_keep, t])
+  }
+
+
+  t <- seq_along(step)
+  expected_infections_vacc <- vapply(t, calculate_expected_infections,
+                                     numeric(dim(S)[1]), no_vacc = FALSE)
+  expected_infections_no_vacc <- vapply(t, calculate_expected_infections,
+                                        numeric(dim(S)[1]), no_vacc = TRUE)
+
+  IFR_t_general <- vnapply(t, calculate_weighted_ratio,
```

Comment:
You might look here at the update that I added for this problem in the Rt calculation. Basically define a list of the options that you're changing (here drop_carehomes, no_vacc and IFR) then loop over the options. That will help with the duplication - see https://github.com/mrc-ide/sircovid/pull/208/files (you don't need to expose the "type" interface though you might choose to)